keg_bouncer/db/mixins.py

```diff
@@ -0,0 +1,76 @@
+import sqlalchemy.orm as saorm
+
+import flask_user
+
+from . import (
+    UserGroup, joined_permission_query, make_link,
+    user_group_permission_map, user_group_bundle_map
+)
+
+
+class UserMixin(flask_user.UserMixin):
+    """A mixin that adds permission facilities to a SQLAlchemy declarative user entity.
+
+    A class which mixes this in must provide the following:
+        * An 'id' field which represents the primary key. It may have any type. If you do not
+          want to call it `id`, you can provide an `id` field as `synonym_for` for your actual
+          primary key.
+    """
+
+    # Instances will override this when caching is needed.
+    _cached_permissions = None
+
+    @declared_attr
+    def id(cls):
+        raise NotImplementedError('You must create an "id" field or a synonym_for it.')
+
+    @declared_attr
+    def user_groups(cls):
+        return saorm.relationship(UserGroup,
+                                  secondary=cls.user_user_group_map,
+                                  cascade='all',
+                                  passive_deletes=True,
+                                  backref='users')
+
+    @declared_attr
+    def user_user_group_map(cls):
+        """A linking (mapping) table between users and user groups."""
+        return make_link('user_user_group_map',
+                         'user_id', cls.id,
+                         'user_group_id', UserGroup.id)
+
+    def get_all_permissions(self):
+        """Get all permissions that are joined to this User, whether directly, through
+           permission bundles, or through user groups."""
+        return frozenset(joined_permission_query().join(
+            self.user_user_group_map,
+            sa.or_(
+                self.user_user_group_map.c.user_group_id
+                    == user_group_permission_map.c.user_group_id,  # noqa
+                self.user_user_group_map.c.user_group_id
+                    == user_group_bundle_map.c.user_group_id  # noqa
+            )
+        ).filter(
+            self.user_user_group_map.c.user_id == self.id,
+        ))
+
+    def get_all_permissions_cached(self):
+        """Same as `get_all_permissions` but caches the result and always returns
+        the same result for this instance of the class.
+
+        This is useful when you want to maximize speed and minimize hits to the database."""
+        if self._cached_permissions is None:
+            self._cached_permissions = self.get_all_permissions()
+        return self._cached_permissions
+
+    def has_permission_token(self, token):
+        """Returns True IFF the given token is present in the user's permission set."""
+        return token in [x.token for x in self.get_all_permissions()]
+
+    def has_permission_token_cached(self, token):
+        """Same as `has_permission_token` except operates on the permission cache."""
+        return token in [x.token for x in self.get_all_permissions()]
```

Comment:
Oops! I thought I had a test case for that...

I'm fine with always using a cache. As far as expiring the cache, I'd rather just have the user code build a new `User` object instead of mutating this one.